---
title: "HIV Genome Graph"
output: 
  flexdashboard::flex_dashboard:
    orientation: columns
    social: menu
    source_code: embed
runtime: shiny
---


```{r global, include=FALSE}
library(tidyverse)
library(ggraph)
library(igraph)
library(rgexf)
#library(plotly)
knitr::opts_chunk$set(message = FALSE)


# Requests:
#   - make edge length == distance between the kmers in the HIV genome, to track linear genome as much as possible
#   - get more attributes (especially genes)

```
```{r preprocess}
### Saria's summary of SWIGG output (error?! what is the first column?)
#anno_output <- "https://raw.githubusercontent.com/NCBI-Codeathons/Virus_Graphs/master/data/HIV_full_Refs_k23_1.tsv.fasta.reference_hits"
anno_output <- "HIV_full_Refs_k23_1.tsv.fasta.reference_hits"
kmer_hits <- read.delim(anno_output, header = FALSE, stringsAsFactors = FALSE)
colnames(kmer_hits) <- c("unknown", "acc_id", "kmer_seq")
kmer_anno <- kmer_hits %>%
  group_by(kmer_seq) %>%
  summarise(
    n_acc_ids = length(unique(acc_id)),
    the_acc_ids = paste(unique(acc_id), collapse = ";")
  ) 

### Load SWIGG output
#tsv_output <- "https://raw.githubusercontent.com/NCBI-Codeathons/Virus_Graphs/master/data/HIV_full_Refs_k23_1.tsv"
tsv_output <- "HIV_full_Refs_k23_1.tsv"
swigg <- read.delim(tsv_output, header = FALSE, stringsAsFactors = FALSE)

### Edge annotation
edges <- swigg
edges$acc_ids <- paste(
  ifelse(
    kmer_anno$the_acc_ids[kmer_anno$kmer_seq == edges$V2]
    ),
  kmer_anno$the_acc_ids[kmer_anno$kmer_seq == edges$V3], 
  collapse = ";"
)
edges_anno1a <- merge(
  edges, kmer_anno, by.x = "V2", by.y = "kmer_seq", all.x = TRUE
)$the_acc_ids
edges_anno1b <- merge(
  edges, kmer_anno, by.x = "V3", by.y = "kmer_seq", all.x = TRUE
)$the_acc_ids
edges$the_acc_ids <- paste(edges_anno1a, edges_anno1b, sep = ";")

edges_anno2a <- merge(edges, kmer_anno, by.x = "V2", by.y = "kmer_seq", all.x = TRUE)$n_acc_ids
edges_anno2b <- merge(edges, kmer_anno, by.x = "V3", by.y = "kmer_seq", all.x = TRUE)$n_acc_ids
edges$n_acc_ids <- edges_anno2a + edges_anno2b

### Node annotation
# use kmer_anno

# Load gfex RDS (from above)
gfex <- "HIV_full_Refs_k23_1_Color_Annotated.gexf.RDS"
#gfex <- "https://github.com/NCBI-Codeathons/Virus_Graphs/blob/master/data/HIV_full_Refs_k23_1_Color_Annotated.gexf.RDS" # not working
#gfex <- "https://raw.githubusercontent.com/NCBI-Codeathons/Virus_Graphs/master/data/HIV_full_Refs_k23_1_Color_Annotated.gexf.RDS" # not working
gr <- readRDS(gfex)

# Add annotation - Edges
E(gr)$n_acc_ids <- edges$n_acc_ids
E(gr)$the_acc_ids <- edges$the_acc_ids

# Add annotation - Nodes
kmer_anno <- kmer_anno[order(kmer_anno$kmer_seq, names(V(gr))), ]
V(gr)$n_acc_ids <- kmer_anno$n_acc_ids
V(gr)$the_acc_ids <- kmer_anno$the_acc_ids
```

Inputs {.sidebar data-width=300}
-----------------------------------------------------------------------

```{r}
### Color by specific accession
selectInput(
  "accA",
  "fasta A",
  sort(unique(kmer_hits$acc_id)),
  selected = sort(unique(kmer_hits$acc_id))[1]
)
selectInput(
  "accB",
  "fasta B",
  sort(unique(kmer_hits$acc_id)),
  selected = sort(unique(kmer_hits$acc_id))[2]
)
selectInput(
  "accC",
  "fasta C",
  sort(unique(kmer_hits$acc_id)),
  selected = sort(unique(kmer_hits$acc_id))[3]
)
selectInput(
  "accD",
  "fasta D",
  sort(unique(kmer_hits$acc_id)),
  selected = sort(unique(kmer_hits$acc_id))[4]
)

### layout algorithm (from igraph)
selectInput(
  "alg",
  "algorithm",
  c("kk", "dh", "nicely", "dh", "fr", "gem", 
    # "grid, "randomly", "sphere", "graphopt", # these aren't useful
    "lgl", "mds", "sugiyama"
  ),
  selected = "kk"
)

### node formatting (not done yet)

### edge formatting (not done yet)

```

Column {}
------------------------------------------------------------------------------

### A

```{r}
renderPlot({
  
  # group 
  E(gr)$accA <- ifelse(
    E(gr)$the_acc_ids %in% 
      E(gr)$the_acc_ids[grep(input$accA, E(gr)$the_acc_ids)], 
    "YES", "NO"
  )
  
  # plot
  set.seed(1116)
  print(
    
    gr %>%
      ggraph(layout = "igraph", algorithm = input$alg) +
      
      # add nodes
      geom_node_point(aes(alpha = n_acc_ids), shape = 21, fill = "black", color = "black", size = 3) +
      
      # add edges
      geom_edge_link(aes(color = accA, alpha = accA, width = accA)) +
      
      # format edges
      scale_edge_width_manual(values = c("YES" = 0.9, "NO" = 0.3)) +
      scale_edge_color_manual(values = c("YES" = "blue", "NO" = "black")) +
      scale_edge_alpha_manual(values = c("YES" = 1, "NO" = 1))
    
    )
})
```

### A vs B

```{r}
renderPlot({
  
  # group 
  E(gr)$accA <- ifelse(
    E(gr)$the_acc_ids %in% 
      E(gr)$the_acc_ids[grep(input$accA, E(gr)$the_acc_ids)], 
    "A", ""
  )
  E(gr)$accB <- ifelse(
    E(gr)$the_acc_ids %in% 
      E(gr)$the_acc_ids[grep(input$accB, E(gr)$the_acc_ids)], 
    "B", ""
  )
  E(gr)$acc <- paste0(E(gr)$accA, E(gr)$accB)
  E(gr)$acc[E(gr)$acc == ""] <- "neither"
    
  # plot
  set.seed(1116)
  print(
    
    gr %>%
      ggraph(layout = "igraph", algorithm = input$alg) +
      
      # add nodes
      geom_node_point(aes(alpha = n_acc_ids), shape = 21, fill = "black", color = "black", size = 3) +
      
      # add edges
      geom_edge_link(aes(color = acc, width = acc)) +
      
      # format edges
      scale_edge_width_manual(values = c("AB" = 0.9, "A" = 0.9, "B" = 0.9, "neither" = 0.3)) +
      scale_edge_color_manual(values = c("AB" = "red", "A" = "blue", "B" = "green", "neither" = "grey25")) 
    
    )
})
```

Column {}
------------------------------------------------------------------------------

### A vs C

```{r}
renderPlot({
  
  # group 
  E(gr)$accA <- ifelse(
    E(gr)$the_acc_ids %in% 
      E(gr)$the_acc_ids[grep(input$accA, E(gr)$the_acc_ids)], 
    "A", ""
  )
  E(gr)$accC <- ifelse(
    E(gr)$the_acc_ids %in% 
      E(gr)$the_acc_ids[grep(input$accC, E(gr)$the_acc_ids)], 
    "C", ""
  )
  E(gr)$acc <- paste0(E(gr)$accA, E(gr)$accC)
  E(gr)$acc[E(gr)$acc == ""] <- "neither"
    
  # plot
  set.seed(1116)
  print(
    
    gr %>%
      ggraph(layout = "igraph", algorithm = input$alg) +
      
      # add nodes
      geom_node_point(aes(alpha = n_acc_ids), shape = 21, fill = "black", color = "black", size = 3) +
      
      # add edges
      geom_edge_link(aes(color = acc, width = acc)) +
      
      # format edges
      scale_edge_width_manual(values = c("AC" = 0.9, "A" = 0.9, "C" = 0.9, "neither" = 0.3)) +
      scale_edge_color_manual(values = c("AC" = "red", "A" = "blue", "C" = "green", "neither" = "grey25")) 
    
    )
})
```

### A vs D

```{r}
renderPlot({
  
  # group 
  E(gr)$accA <- ifelse(
    E(gr)$the_acc_ids %in% 
      E(gr)$the_acc_ids[grep(input$accA, E(gr)$the_acc_ids)], 
    "A", ""
  )
  E(gr)$accD <- ifelse(
    E(gr)$the_acc_ids %in% 
      E(gr)$the_acc_ids[grep(input$accD, E(gr)$the_acc_ids)], 
    "D", ""
  )
  E(gr)$acc <- paste0(E(gr)$accA, E(gr)$accD)
  E(gr)$acc[E(gr)$acc == ""] <- "neither"
    
  # plot
  set.seed(1116)
  print(
    
    gr %>%
      ggraph(layout = "igraph", algorithm = input$alg) +
      
      # add nodes
      geom_node_point(aes(alpha = n_acc_ids), shape = 21, fill = "black", color = "black", size = 3) +
      
      # add edges
      geom_edge_link(aes(color = acc, width = acc)) +
      
      # format edges
      scale_edge_width_manual(values = c("AD" = 0.9, "A" = 0.9, "D" = 0.9, "neither" = 0.3)) +
      scale_edge_color_manual(values = c("AD" = "red", "A" = "blue", "D" = "green", "neither" = "grey25")) 
    
    )
})
```
